README.md
# Kinvey Datalink Errors

Create Kinvey-formatted errors for Datalinks

## Installation

```
npm install kinvey-datalink-errors
```

## Usage

Kinvey Errors exposes a single function, `generateKinveyError` which takes the error type and an error.  The error can
either be a string or a JavaScript `Error`.

```
var kinveyErrors = require('kinvey-datalink-errors');

var myError = kinveyErrors.generateKinveyError("NotFound", "The record is not found"); // a string
var myError = kinveyErrors.generateKinveyError("NotFound, new Error("The record is not found")); // an error
```

The library returns the error passed as a Kinvey-formatted error.

```
{
  error: "NotFound",
  description: "The requested entity or entites were not found in the collection."
  debug: "THe record is not found"
  statusCode: 404
}

If an error object is passed and it contains a stack trace, the stack trace will also be included in the error object.

## Error Types

This library exposes the following error types:


| Error Type    | Description  |
|-------------|-------------|
|NotFound      | Used when an entity is not found.  Note:  This should only be used on getById and deleteById commands.  Otherwise, an empty array should be returned (for getAll, getByQuery) or a count of 0 (for deleteByQuery, countByQuery).|
| BadRequest    | Used for any invalid input - queries, body, etc. |
| Unauthorized  | Used when the operation or user is unauthorized to perform the operation.  |
| Forbidden     | Used when an operation is forbidden. |
| NotAllowed    | Used when the request is not allowed |
| NotImplemented | Used when a specific method is not implemented |
| RuntimeError | Used for any type of runtime error |


CHANGELOG.md
## Changelog

### 0.3.4
* Fixed several security vulnerabilities

### 0.3.3
* Change build of node.js version in travis
* Fixed linting
* Published to npm
* Fixed inheritence error
* Update dependency versions

### 0.3.2
* Removed coffeescript register in index.js

### v0.3.1
* Converted to javascript

### v0.3.0
* Add toJSON method to KinveyError

### v0.2.2
* Updated copyrights and licenses

### v0.2.1
* Fixed NotFound error to be a 404 rather than a 400 error

### v0.2.0
* Created KinveyError as a subclass of Error

### v0.1.0
* Initail Release


test/unit/lib/errors.test.js
// Copyright (c) 2018 Kinvey Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

const errors = require('../../../lib/errors');
const should = require('should');

describe('process error debug messages', () => {
  it('should process an error when a string is passed', () => {
    const myError = errors.generateKinveyError('NotFound', 'This is a test error');
    myError.debug.should.eql('This is a test error');
  });

  it('should process an error when an error object is passed', () => {
    const myError = errors.generateKinveyError('NotFound', new Error('This is a test error'));
    myError.debug.should.eql('This is a test error');
    should.exist(myError.stack);
  });

  it('should process an error when nothing is passed', () => {
    const myError = errors.generateKinveyError('NotFound');
    myError.debug.should.eql({});
  });

  it('should output an error as raw json', () => {
    const myError = errors.generateKinveyError('NotFound', 'this is a test error');
    const errObj = myError.toJSON();
    errObj.should.not.be.instanceOf(Error);
    errObj.should.be.instanceOf(Object);
    errObj.error.should.eql('NotFound');
    errObj.description.should.eql('The requested entity or entites were not found in the collection');
    errObj.debug.should.eql('this is a test error');
    errObj.statusCode.should.eql(404);
    should.not.exist(errObj.stack);
  });

  it('should output an error as raw json with a stack trace', () => {
    const myError = errors.generateKinveyError('NotFound', new Error('this is a test error'));
    const errObj = myError.toJSON();
    errObj.should.not.be.instanceOf(Error);
    errObj.should.be.instanceOf(Object);
    errObj.error.should.eql('NotFound');
    errObj.description.should.eql('The requested entity or entites were not found in the collection');
    errObj.debug.should.eql('this is a test error');
    errObj.statusCode.should.eql(404);
    should.exist(errObj.stack);
  });
});

describe('process errors of different types', () => {
  it('should process a NotFound Error', () => {
    const myError = errors.generateKinveyError('NotFound', 'Debug Text');
    myError.message.should.eql('NotFound');
    myError.description.should.eql('The requested entity or entites were not found in the collection');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(404);
  });

  it('should process a BadRequest Error', () => {
    const myError = errors.generateKinveyError('BadRequest', 'Debug Text');
    myError.message.should.eql('BadRequest');
    myError.description.should.eql('Unable to understand the request');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(400);
  });

  it('should process an InvalidCredentials Error', () => {
    const myError = errors.generateKinveyError('Unauthorized', 'Debug Text');
    myError.message.should.eql('InvalidCredentials');
    myError.description.should.eql('Invalid credentials.  Please retry your request with correct credentials.');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(401);
  });

  it('should process a Forbidden Error', () => {
    const myError = errors.generateKinveyError('Forbidden', 'Debug Text');
    myError.message.should.eql('Forbidden');
    myError.description.should.eql('The request is forbidden');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(403);
  });

  it('should process a NotAllowed Error', () => {
    const myError = errors.generateKinveyError('NotAllowed', 'Debug Text');
    myError.message.should.eql('NotAllowed');
    myError.description.should.eql('The request is not allowed');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(405);
  });

  it('should process a NotImplemented Error', () => {
    const myError = errors.generateKinveyError('NotImplemented', 'Debug Text');
    myError.message.should.eql('NotImplemented');
    myError.description.should.eql('The request invoked a method that is not implemented');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(501);
  });

  it('should process a Runtime Error', () => {
    const myError = errors.generateKinveyError('RuntimeError', 'Debug Text');
    myError.message.should.eql('DataLinkRuntimeError');
    myError.description.should.eql('The datalink had a runtime error.  See debug message for details');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(550);
  });

  it('should return a default error if the error is unknown', () => {
    const myError = errors.generateKinveyError('SomeRandomError', 'Debug Text');
    myError.message.should.eql('DataLinkInternalError');
    myError.description.should.eql('The DataLink request experienced a problem. See debug message for details.');
    myError.debug.should.eql('Debug Text');
    myError.statusCode.should.eql(550);
  });
});



lib/errors.js
// Copyright (c) 2018 Kinvey Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

const util = require('util');

const errTypes = {
  NotFound: {
    error: 'NotFound',
    description: 'The requested entity or entites were not found in the collection',
    statusCode: 404
  },
  BadRequest: {
    error: 'BadRequest',
    description: 'Unable to understand the request',
    statusCode: 400
  },
  Unauthorized: {
    error: 'InvalidCredentials',
    description: 'Invalid credentials.  Please retry your request with correct credentials.',
    statusCode: 401
  },
  Forbidden: {
    error: 'Forbidden',
    description: 'The request is forbidden',
    statusCode: 403
  },
  NotAllowed: {
    error: 'NotAllowed',
    description: 'The request is not allowed',
    statusCode: 405
  },
  NotImplemented: {
    error: 'NotImplemented',
    description: 'The request invoked a method that is not implemented',
    statusCode: 501
  },
  RuntimeError: {
    error: 'DataLinkRuntimeError',
    description: 'The datalink had a runtime error.  See debug message for details',
    statusCode: 550
  },
  IncorrectContextRoot: {
    error: 'IncorrectContextRoot',
    description: 'Incorrect context root specified. See debug message for details',
    statusCode: 550
  },
  default: {
    error: 'DataLinkInternalError',
    description: 'The DataLink request experienced a problem. See debug message for details.',
    statusCode: 550
  }
};

class KinveyError {
  constructor(message) {
    Error.call(this);
    this.message = message;
  }
  toJSON() {
    const jsonError = {
      error: this.message,
      description: this.description,
      debug: this.debug
    };

    if (this.statusCode) {
      jsonError.statusCode = this.statusCode;
    }

    if (this.stack) {
      jsonError.stack = this.stack;
    }

    return jsonError;
  }
}

util.inherits(KinveyError, Error);

function generateKinveyError(type, err) {
  const error = errTypes[type] || errTypes.default;
  const kinveyError = new KinveyError(error.error);
  kinveyError.description = error.description;
  kinveyError.statusCode = error.statusCode;

  if (err && err instanceof Error) {
    kinveyError.debug = err.debug || err.message || err.name || err.toString() || {};

    if (err.stack) {
      kinveyError.stack = err.stack;
    }
  } else {
    kinveyError.debug = err || {};
  }

  return kinveyError;
}

exports.generateKinveyError = generateKinveyError;
